Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,609 @@
+            }
+            self._moveHead(to: 0)
+            self._moveTail(to: w - r)
+            
+        case let (r, w) where self._storage.capacity - r <= r - w:
+            // reader is above writer, but we have room in the middle to move the lower
+            // segment forward to make way for the higher segment
+            // i.e.:
+            //      +---------------------------------------+
+            //  1.  |------W                        R-------|
+            //      +---------------------------------------+
+            //
+            //      +---------------------------------------+
+            //  2.  |        ------W                R-------|
+            //      +---------------------------------------+
+            //
+            //      +---------------------------------------+
+            //  3.  |R-------------W                        |
+            //      +---------------------------------------+
+            self._storage.withVeryUnsafeBytes { ptr in
+                let writePtr = UnsafeMutableRawPointer(mutating: ptr.baseAddress!)
+                let endSrc = UnsafeRawBufferPointer(start: ptr.baseAddress!, count: w)
+                let startSrc = UnsafeRawBufferPointer(start: ptr.baseAddress!.advanced(by: r), count: ptr.count - r)
+                
+                writePtr.advanced(by: startSrc.count).copyMemory(from: endSrc.baseAddress!, byteCount: endSrc.count)
+                writePtr.copyMemory(from: startSrc.baseAddress!, byteCount: startSrc.count)
+            }
+            self._moveHead(to: 0)
+            self._moveTail(to: _storage.capacity - r + w)
+            
+        case let (r, w):
+            // every other option has been exhausted. Now we have a buffer
+            // where reader > writer, and there's not enough room in the middle
+            // to move things around. Now while we could look at the available
+            // space and shuffle that many bytes around, it's likely no faster
+            // than reallocating and copying.
+            var newBytes = self._storage
+            newBytes.clear()
+            newBytes.write(bytes: self._storage.viewBytes(at: r, length: self._storage.capacity - r))
+            newBytes.write(bytes: self._storage.viewBytes(at: 0, length: w))
+            self._storage = newBytes
+            self._moveHead(to: 0)
+            self._moveTail(to: _storage.capacity - r + w)
+        }
+    }
+    
+    @inlinable
+    mutating func _moveHead(to newIndex: Int) {
+        assert(newIndex >= 0 && newIndex <= self._storage.capacity)
+        self._ringHead = newIndex
+    }
+    
+    @inlinable
+    mutating func _moveHead(forwardBy amount: Int) {
+        assert(amount >= 0)
+        let readerIsBelowWriter = self._ringHead <= self._ringTail
+        var newIndex = self._ringHead + amount
+        
+        // reader index should not pass writer index
+        if readerIsBelowWriter {
+            assert(newIndex <= self._ringTail)
```

Comment:
These asserts look like they should be preconditions to me, as they affect the correctness of the buffer if they're mishandled.